python/ray/autoscaler/v2/instance_manager/subscribers/threaded_ray_installer.py

```diff
@@ -0,0 +1,102 @@
+import logging
+import time
+from concurrent.futures import ThreadPoolExecutor
+from typing import List
+
+from ray.autoscaler.v2.instance_manager.instance_storage import (
+    InstanceStorage,
+    InstanceUpdatedSuscriber,
+    InstanceUpdateEvent,
+)
+from ray.autoscaler.v2.instance_manager.ray_installer import RayInstaller
+from ray.core.generated.instance_manager_pb2 import Instance
+
+logger = logging.getLogger(__name__)
+
+
+class ThreadedRayInstaller(InstanceUpdatedSuscriber):
+    """ThreadedRayInstaller is responsible for install ray on new nodes."""
+
+    def __init__(
+        self,
+        head_node_ip: str,
+        instance_storage: InstanceStorage,
+        ray_installer: RayInstaller,
+        max_install_attempts: int = 3,
+        install_retry_interval: int = 10,
+        max_concurrent_installs: int = 50,
+    ) -> None:
+        self._head_node_ip = head_node_ip
+        self._instance_storage = instance_storage
+        self._ray_installer = ray_installer
+        self._max_concurrent_installs = max_concurrent_installs
+        self._max_install_attempts = max_install_attempts
+        self._install_retry_interval = install_retry_interval
+        self._ray_installation_executor = ThreadPoolExecutor(
+            max_workers=self._max_concurrent_installs
+        )
+
+    def notify(self, events: List[InstanceUpdateEvent]) -> None:
+        for event in events:
+            if (
+                event.new_status == Instance.ALLOCATED
+                and event.new_ray_status == Instance.RAY_STATUS_UNKOWN
+            ):
+                self._install_ray_on_new_nodes(event.instance_id)
```

Comment:
should this be posted to another thread executor as well?

Reply:
probably fine. we just pay a small instance load overhead.